File: src/lib/services/docker/swarm.ts
import { config } from "$lib/config";
import { Logger } from "$lib/logger";
import type { PublishedPort } from "$lib/published-ports";
import type { ContainerStatus } from "$lib/types";
import type { BaseDockerService, Constructor } from "./base.ts";
import type { RegistryAuth, VolumeMountParams } from "./containers.ts";
import { buildContainerLabels, SERVICE_ID_LABEL } from "./labels.ts";
import {
	type ContainerRuntimeParams,
	capabilityName,
	mergeLabels,
} from "./runtime-options.ts";

const logger = new Logger("Swarm");

export const SWARM_REFRESH_SECONDS = 2;

/**
 * Swarm services can only join an **overlay** network, and the shared
 * network standalone containers use (`config.docker.networkName`) already
 * exists as a bridge on any instance that ever deployed anything, with
 * Traefik and the app itself attached to it. A live bridge network can't be
 * converted, so swarm mode gets its own attachable overlay alongside it
 * rather than fighting over one name : Traefik joins both and routes either
 * way.
 */
export function swarmNetworkName(): string {
	return `${config.docker.networkName}-swarm`;
}

/**
 * Whether `docker info`'s security options mark the daemon as rootless.
 * Swarm mode can't work there: a rootless daemon can't create the overlay
 * network a swarm service joins, and fails the attach with a bare "context
 * deadline exceeded" after the swarm was already initialised.
 */
export function isRootlessDaemon(
	securityOptions: string[] | undefined,
): boolean {
	return (securityOptions ?? []).some((option) =>
		option.split(",").includes("name=rootless"),
	);
}

/**
 * Why swarm mode can't run on the daemon `docker info` describes, or null
 * when it can. A rootless daemon can't create the overlay network a swarm
 * service joins.
 */
export function swarmUnavailableReason(
	securityOptions: string[] | undefined,
): string | null {
	if (!isRootlessDaemon(securityOptions)) {
		return null;
	}
	return "This Docker daemon runs rootless, and rootless Docker can't create the overlay networks swarm services need. Move Homerun onto the system (rootful) daemon first: sudo homerun-installer --migrate-to-rootful.";
}

/**
 * The orchestration mode a brand new instance starts in: swarm when this
 * daemon is already a swarm manager and isn't rootless (what the default
 * installer sets up), standalone otherwise.
 */
export function initialOrchestrationMode(info: {
	SecurityOptions?: string[];
	Swarm?: { ControlAvailable?: boolean; LocalNodeState?: string };
}): "standalone" | "swarm" {
	const manager =
		info.Swarm?.LocalNodeState === "active" &&
		info.Swarm.ControlAvailable === true;
	return manager && swarmUnavailableReason(info.SecurityOptions) === null
		? "swarm"
		: "standalone";
}

/**
 * A service volume as a swarm mount: an absolute source is a bind mount of
 * that host path, anything else is a named volume, the same rule Docker's
 * `Binds` syntax applies to a standalone container.
 */
export function swarmMount(volume: VolumeMountParams): {
	ReadOnly: boolean;
	Source: string;
	Target: string;
	Type: "bind" | "volume";
} {
	return {
		ReadOnly: volume.readOnly,
		Source: volume.source,
		Target: volume.containerPath,
		Type: volume.source.startsWith("/") ? "bind" : "volume",
	};
}

/**
 * The networks a swarm service's tasks attach to: the host's own network
 * stack for host networking, otherwise the shared overlay with the slug as
 * an alias, so other services reach it at `http://<slug>:<port>` the same
 * way they reach a standalone container.
 */
export function swarmNetworksFor(
	networkMode: "bridge" | "host" | undefined,
	overlay: string,
	slug: string,
): { Aliases?: string[]; Target: string }[] {
	if (networkMode === "host") {
		return [{ Target: "host" }];
	}
	return [{ Aliases: [slug], Target: overlay }];
}

/** Swarm's restart conditions for this app's restart policy values: "no" never restarts, "on-failure" only after a non-zero exit, anything else always. */
export function swarmRestartCondition(
	restartPolicy: string,
): "any" | "none" | "on-failure" {
	if (restartPolicy === "no") {
		return "none";
	}
	return restartPolicy === "on-failure" ? "on-failure" : "any";
}

/**
 * The swarm service spec for a service, minus its name and what depends on
 * the image the deploy actually resolved (the task's `Env`, `Image`, readiness
 * `Healthcheck` and label): Traefik and tracking labels on the service (routed
 * over the swarm overlay), the replica count, and a task template with the
 * runtime options, mounts (`swarmMount`), networks (`swarmNetworksFor`),
 * resource limits and restart condition. The homerun worker fills in the rest.
 */
export function swarmServiceTemplate(params: CreateSwarmServiceParams) {
	return {
		Labels: mergeLabels(
			params.runtime?.labels,
			buildContainerLabels({
				containerPort: params.containerPort,
				defaultDomainEnabled: params.defaultDomainEnabled,
				dnsResolvable:
					params.networkMode === "host" ? false : params.dnsResolvable,
				domainPorts: params.domainPorts,
				domains: params.domains,
				networkName: swarmNetworkName(),
				serviceId: params.serviceId,
				slug: params.slug,
				stackSlug: params.stackSlug,
			}),
		),
		EndpointSpec: params.publishedPorts?.length
			? {
					Ports: params.publishedPorts.map((port) => ({
						Protocol: port.protocol,
						PublishedPort: port.hostPort,
						PublishMode: "ingress",
						TargetPort: port.containerPort,
					})),
				}
			: undefined,
		Mode: { Replicated: { Replicas: params.replicas } },
		TaskTemplate: {
			ContainerSpec: {
				Args: params.runtime?.command ?? undefined,
				CapabilityAdd: params.runtime?.capAdd.length
					? params.runtime.capAdd.map(capabilityName)
					: undefined,
				Command: params.runtime?.entrypoint ?? undefined,
				Labels: mergeLabels(params.runtime?.labels, {
					[SERVICE_ID_LABEL]: params.serviceId,
					"homerun.managed": "true",
				}),
				Mounts: (params.volumes ?? []).map(swarmMount),
			},
			Networks: swarmNetworksFor(
				params.networkMode,
				swarmNetworkName(),
				params.slug,
			),
			Resources: {
				Limits: {
					MemoryBytes: params.memoryLimitMb
						? params.memoryLimitMb * 1024 * 1024
						: undefined,
					NanoCPUs: params.cpuLimit
						? Math.round(Number.parseFloat(params.cpuLimit) * 1e9)
						: undefined,
				},
			},
			RestartPolicy: {
				Condition: swarmRestartCondition(params.restartPolicy),
			},
		},
	};
}

interface DaemonInfo {
	SecurityOptions?: string[];
	ServerVersion?: string;
	Swarm?: {
		ControlAvailable?: boolean;
		LocalNodeState?: string;
		NodeID?: string;
	};
}

export interface SwarmTask {
	DesiredState?: string;
	ID?: string;
	NodeID?: string;
	ServiceID?: string;
	Slot?: number;
	Status?: {
		ContainerStatus?: { ContainerID?: string; ExitCode?: number } | null;
		Err?: string;
		Message?: string;
		State?: string;
		Timestamp?: string;
	};
	UpdatedAt?: string | null;
}

export interface SwarmReadiness {
	network: string;
	overlayReady: boolean;
	swarmActive: boolean;
	traefikFound: boolean;
	traefikSwarmProvider: boolean;
}

export interface CreateSwarmServiceParams {
	auth?: RegistryAuth;
	containerPort: number;
	cpuLimit?: string | null;
	defaultDomainEnabled?: boolean;
	dnsResolvable?: boolean;
	domainPorts?: Record<string, number>;
	domains?: string[];
	envVars: Record<string, string>;
	image: string;
	memoryLimitMb?: number | null;
	networkMode?: "bridge" | "host";
	portProtocol?: "tcp" | "udp" | "both";
	publishedPorts?: PublishedPort[];
	healthcheckCommand?: string | null;
	runtime?: ContainerRuntimeParams;
	stackSlug?: string | null;
	replicas: number;
	restartPolicy: string;
	serviceId: string;
	slug: string;
	tag: string;
	volumes?: VolumeMountParams[];
}

/**
 * Docker Swarm equivalent of docker/containers.ts, for services deployed
 * under `instanceSettings.orchestrationMode === "swarm"` : a swarm
 *Service* (replicated, self-healing, scalable) instead of a single
 * container. Deliberately local-only for v1 : this app's Remote Hosts
 * feature (an arbitrary separate Docker daemon) doesn't apply the same way
 * under swarm, a "remote" node has to actually join *this* swarm as a
 * worker and get targeted via placement constraints, not connected to as
 * an independent client. Not built here, flagged as a real gap : swarm-mode
 * services always run against the local swarm manager.
 *
 * The host is prepared by `enableSwarmMode` (core-services.ts) when the
 * orchestration mode is switched on the dashboard : `docker swarm init`
 * through `initSwarm` below, the overlay network, and Traefik's
 * `--providers.swarm` flags on the live container.
 */
// oxlint-disable-next-line max-lines-per-function -- mixin factory: the body is a class definition, not a procedure
export function DockerSwarmMixin<TBase extends Constructor<BaseDockerService>>(
	Base: TBase,
) {
	return class DockerSwarmService extends Base {
		/** Whether this daemon is already a swarm manager. */
		async isSwarmActive(): Promise<boolean> {
			const info = await this.worker.get<DaemonInfo>("/v1/info");
			return info?.Swarm?.LocalNodeState === "active";
		}

		/**
		 * Refuses to go further on a rootless daemon, before anything on the
		 * host has been changed.
		 *
		 * @throws When this daemon runs rootless.
		 */
		async assertSwarmCapableDaemon(): Promise<void> {
			const info = await this.worker.get<DaemonInfo>("/v1/info");
			const reason = swarmUnavailableReason(info?.SecurityOptions);
			if (reason) {
				throw new Error(reason);
			}
		}

		/**
		 * Why swarm mode can't be switched on for this daemon, for the settings
		 * page to show before anyone submits. Null when it can, and also when
		 * the daemon can't be reached, since that isn't a reason to refuse.
		 */
		async swarmModeUnavailableReason(): Promise<string | null> {
			const info = await this.worker
				.get<DaemonInfo>("/v1/info")
				.catch(() => null);
			return info ? swarmUnavailableReason(info.SecurityOptions) : null;
		}

		/**
		 * The mode a fresh instance should start in, from the live daemon (see
		 * `initialOrchestrationMode`). Standalone when the daemon can't be
		 * reached.
		 */
		async detectInitialOrchestrationMode(): Promise<"standalone" | "swarm"> {
			const info = await this.worker
				.get<DaemonInfo>("/v1/info")
				.catch(() => null);
			return info ? initialOrchestrationMode(info) : "standalone";
		}

		/**
		 * `docker swarm init` on this host, idempotent : returns false when
		 * the daemon was already a manager. The one place this app runs a
		 * host-level daemon operation, and it does so only because the
		 * dashboard's own Swarm switch is otherwise a setting that changes
		 * nothing.
		 */
		async initSwarm(): Promise<boolean> {
			const { alreadyActive } = await this.worker.post<{
				alreadyActive: boolean;
			}>("/v1/swarm/init");
			if (alreadyActive) {
				return false;
			}
			logger.info("Swarm initialised on this host");
			return true;
		}

		/**
		 * What swarm mode needs before a service can be deployed under it,
		 * each checked against the live daemon rather than assumed : the
		 * daemon being a swarm manager, the overlay network existing, and
		 * Traefik actually running its swarm provider. Read-only.
		 */
		async swarmReadiness(): Promise<SwarmReadiness> {
			const network = swarmNetworkName();
			const [active, networks] = await Promise.all([
				this.isSwarmActive().catch(() => false),
				this.worker
					.get<{ Name: string }[]>("/v1/networks")
					.catch(() => [] as { Name: string }[]),
			]);
			const traefik = await this.worker
				.get<{ Id: string; Image: string }[]>("/v1/containers", { all: "1" })
				.then((rows) => rows.find((row) => row.Image.startsWith("traefik")))
				.catch(() => undefined);
			const cmd: string[] = traefik
				? await this.worker
						.get<{ Config?: { Cmd?: string[] | null } }>(
							`/v1/containers/${traefik.Id}/inspect`,
						)
						.then((info) => info.Config?.Cmd ?? [])
						.catch(() => [])
				: [];
			return {
				network,
				overlayReady: networks.some((net) => net.Name === network),
				swarmActive: active,
				traefikFound: !!traefik,
				traefikSwarmProvider: cmd.some(
					(arg) =>
						arg.startsWith("--providers.swarm=true") ||
						arg.startsWith("--providers.docker.swarmMode=true"),
				),
			};
		}

		/** Idempotent : swarm networks are cluster-wide, created once and reused by every swarm-mode service. */
		async ensureSwarmNetwork(name: string): Promise<void> {
			await this.worker.post<{ ok: boolean }>("/v1/swarm/overlay", { name });
			logger.info(`Swarm overlay network ready: ${name}`);
		}

		/** Removes a swarm service. One already gone isn't an error. */
		async removeSwarmService(swarmServiceId: string): Promise<void> {
			await this.worker.delete<{ ok: boolean }>(
				`/v1/swarm/services/${swarmServiceId}`,
			);
		}

		/** "Stop"/"start" in swarm terms : scale replicas to 0, or back up to the service's configured count. */
		async scaleSwarmService(
			swarmServiceId: string,
			replicas: number,
		): Promise<void> {
			await this.worker.post<{ ok: boolean }>(
				`/v1/swarm/services/${swarmServiceId}/scale`,
				{ replicas },
			);
		}

		/** "Restart" in swarm terms : a force-update, which recreates every task's container even though the spec is unchanged. */
		async restartSwarmService(swarmServiceId: string): Promise<void> {
			await this.worker.post<{ ok: boolean }>(
				`/v1/swarm/services/${swarmServiceId}/restart`,
			);
		}

		/** Aggregate status across every task of the service, same ContainerStatus vocabulary standalone mode uses. */
		async inspectSwarmServiceStatus(
			swarmServiceId: string,
		): Promise<ContainerStatus> {
			const data = await this.worker.get<{ status: ContainerStatus }>(
				`/v1/swarm/services/${swarmServiceId}/status`,
			);
			return data.status;
		}

		/**
		 * The container id of one running task of the service on this node, for
		 * the Terminal tab's exec and pre-backup commands (swarm has no
		 * service-level exec, only container-level, and this app only reaches
		 * the local daemon). Null when every running replica is on another node.
		 */
		async getRunningTaskContainerId(
			swarmServiceId: string,
		): Promise<string | null> {
			const [tasks, info] = await Promise.all([
				this.worker.get<SwarmTask[]>(
					`/v1/swarm/services/${swarmServiceId}/tasks`,
					{ running: "1" },
				),
				this.worker.get<DaemonInfo>("/v1/info"),
			]);
			const localNodeId = info?.Swarm?.NodeID;
			const running = tasks.find(
				(task) =>
					task.Status?.State === "running" &&
					(!localNodeId || task.NodeID === localNodeId),
			);
			return running?.Status?.ContainerStatus?.ContainerID ?? null;
		}

		/** Same Web ReadableStream shape and options as containers.ts's streamLogs, so a logs route doesn't need to know which mode it's in. The worker demuxes the frames, so what comes back is already plain text. */
		streamSwarmServiceLogs(
			swarmServiceId: string,
			opts?: { tail?: number; follow?: boolean },
		): Promise<ReadableStream<Uint8Array>> {
			return this.worker.stream(`/v1/swarm/services/${swarmServiceId}/logs`, {
				query: {
					follow: opts?.follow === false ? "0" : "1",
					tail: String(opts?.tail ?? 200),
				},
			});
		}
	};
}

File: internal/workerapi/swarm.go
package workerapi

import (
	"errors"
	"net/http"

	"github.com/go-chi/chi/v5"

	"github.com/orochibraru/homerun/internal/dockerapi"
	"github.com/orochibraru/homerun/internal/httpapi"
)

// mountSwarm wires the swarm routes.
func (s *Server) mountSwarm(r chi.Router) {
	r.Post("/v1/swarm/init", httpapi.H(s.initSwarm))
	r.Get("/v1/swarm/nodes", httpapi.H(s.listSwarmNodes))
	r.Post("/v1/swarm/overlay", httpapi.H(s.ensureOverlay))
	r.Post("/v1/swarm/services", httpapi.H(s.createSwarmService))
	r.Get("/v1/swarm/services/{id}", httpapi.H(s.inspectSwarmService))
	r.Get("/v1/swarm/services/{id}/status", httpapi.H(s.swarmServiceStatus))
	r.Get("/v1/swarm/services/{id}/replicas", httpapi.H(s.swarmServiceReplicas))
	r.Get("/v1/swarm/services/{id}/tasks", httpapi.H(s.swarmServiceTasks))
	r.Get("/v1/swarm/services/{id}/logs", httpapi.H(s.swarmServiceLogs))
	r.Post("/v1/swarm/services/{id}/scale", httpapi.H(s.scaleSwarmService))
	r.Post("/v1/swarm/services/{id}/restart", httpapi.H(s.restartSwarmService))
	r.Delete("/v1/swarm/services/{id}", httpapi.H(s.removeSwarmService))
}

// initSwarm turns this daemon into a single-node swarm manager.
func (s *Server) initSwarm(w http.ResponseWriter, r *http.Request) error {
	info, err := s.docker.SystemInfo(r.Context())
	if err != nil {
		return err
	}
	if info.Swarm.LocalNodeState == "active" && info.Swarm.ControlAvailable {
		return httpapi.Answer(w, http.StatusOK, map[string]bool{"alreadyActive": true})
	}
	if err := s.docker.InitSwarm(r.Context(), ""); err != nil {
		return err
	}
	return httpapi.Answer(w, http.StatusOK, map[string]bool{"alreadyActive": false})
}

// listSwarmNodes answers with the swarm's nodes, which the replicas panel uses
// to label each replica with the host running it.
func (s *Server) listSwarmNodes(w http.ResponseWriter, r *http.Request) error {
	nodes, err := s.docker.ListSwarmNodes(r.Context())
	if err != nil {
		return err
	}
	if nodes == nil {
		nodes = []dockerapi.SwarmNode{}
	}
	return httpapi.Answer(w, http.StatusOK, nodes)
}

// ensureOverlayBody names the overlay network to create.
type ensureOverlayBody struct {
	Labels map[string]string `json:"labels"`
	Name   string            `json:"name"`
}

// ensureOverlay creates the attachable overlay a swarm-mode host needs
// alongside the shared bridge, treating one that exists as success.
func (s *Server) ensureOverlay(w http.ResponseWriter, r *http.Request) error {
	var input ensureOverlayBody
	if !httpapi.DecodeJSON(w, r, &input) {
		return nil
	}
	if input.Name == "" {
		return httpapi.Invalid(w, "Invalid request body",
			[]httpapi.ValidationIssue{{Message: "name is required", Path: []string{"name"}}})
	}
	if err := s.docker.EnsureOverlayNetwork(r.Context(), input.Name, input.Labels); err != nil {
		return err
	}
	return httpapi.Answer(w, http.StatusOK, map[string]bool{"ok": true})
}

// createSwarmService creates a swarm service from a raw Engine spec, for the
// app's own core services (the Newt tunnel in swarm mode); a deploy's services
// are created by the deploy job instead.
func (s *Server) createSwarmService(w http.ResponseWriter, r *http.Request) error {
	var spec map[string]any
	if !httpapi.DecodeJSON(w, r, &spec) {
		return nil
	}
	if name, _ := spec["Name"].(string); name == "" {
		return httpapi.Invalid(w, "Invalid request body",
			[]httpapi.ValidationIssue{{Message: "Name is required", Path: []string{"Name"}}})
	}
	id, err := s.docker.CreateSwarmService(r.Context(), spec)
	if err != nil {
		return err
	}
	return httpapi.Answer(w, http.StatusCreated, map[string]string{"id": id})
}

// inspectSwarmService answers with a swarm service's id, name and labels,
// looked up by id or name.
func (s *Server) inspectSwarmService(w http.ResponseWriter, r *http.Request) error {
	service, err := s.docker.InspectSwarmService(r.Context(), chi.URLParam(r, "id"))
	if errors.Is(err, dockerapi.ErrNotFound) {
		return httpapi.Error(w, http.StatusNotFound, "No such swarm service.")
	}
	if err != nil {
		return err
	}
	return httpapi.Answer(w, http.StatusOK, service)
}

// swarmServiceStatus aggregates a service's tasks into one dashboard status.
//
// The rule is the TypeScript one, unchanged: a service scaled to zero is
// stopped, one with any running task is running, one with a failed or rejected
// task is failed, one with no tasks yet is pending, and anything else is still
// starting. A service the daemon can't inspect at all reads as stopped, since
// that's what a removed service looks like from the dashboard's side.
func (s *Server) swarmServiceStatus(w http.ResponseWriter, r *http.Request) error {
	id := chi.URLParam(r, "id")
	replicas, err := s.docker.SwarmServiceReplicas(r.Context(), id)
	if err != nil {
		return httpapi.Answer(w, http.StatusOK, map[string]string{"status": "stopped"})
	}
	if replicas == 0 {
		return httpapi.Answer(w, http.StatusOK, map[string]string{"status": "stopped"})
	}
	tasks, err := s.docker.ListSwarmTasks(r.Context(), id, false)
	if err != nil {
		return httpapi.Answer(w, http.StatusOK, map[string]string{"status": "stopped"})
	}
	return httpapi.Answer(w, http.StatusOK, map[string]string{"status": swarmStatusFrom(tasks)})
}

// swarmStatusFrom reduces a service's tasks to one status.
func swarmStatusFrom(tasks []dockerapi.SwarmTask) string {
	if len(tasks) == 0 {
		return "pending"
	}
	failed := false
	for _, task := range tasks {
		switch task.Status.State {
		case "running":
			return "running"
		case "failed", "rejected":
			failed = true
		}
	}
	if failed {
		return "failed"
	}
	return "starting"
}

// swarmServiceReplicas answers with a service's configured replica count.
func (s *Server) swarmServiceReplicas(w http.ResponseWriter, r *http.Request) error {
	replicas, err := s.docker.SwarmServiceReplicas(r.Context(), chi.URLParam(r, "id"))
	if errors.Is(err, dockerapi.ErrNotFound) {
		return httpapi.Error(w, http.StatusNotFound, "No such swarm service.")
	}
	if err != nil {
		return err
	}
	return httpapi.Answer(w, http.StatusOK, map[string]int{"replicas": replicas})
}

// swarmServiceTasks answers with a service's tasks, optionally only the ones
// the daemon wants running (the rest are previous generations swarm keeps).
func (s *Server) swarmServiceTasks(w http.ResponseWriter, r *http.Request) error {
	runningOnly := r.URL.Query().Get("running") == "1"
	tasks, err := s.docker.ListSwarmTasks(r.Context(), chi.URLParam(r, "id"), runningOnly)
	if err != nil {
		return err
	}
	if tasks == nil {
		tasks = []dockerapi.SwarmTask{}
	}
	return httpapi.Answer(w, http.StatusOK, tasks)
}

// swarmServiceLogs streams a swarm service's aggregated task logs.
func (s *Server) swarmServiceLogs(w http.ResponseWriter, r *http.Request) error {
	follow := r.URL.Query().Get("follow") != "0"
	logs, err := s.docker.SwarmServiceLogs(r.Context(), chi.URLParam(r, "id"), tailParam(r), follow)
	if errors.Is(err, dockerapi.ErrNotFound) {
		return httpapi.Error(w, http.StatusNotFound, "No such swarm service.")
	}
	if err != nil {
		return err
	}
	defer func() { _ = logs.Close() }()
	return streamOut(w, r, logs)
}

// scaleSwarmServiceBody is the replica count to scale to.
type scaleSwarmServiceBody struct {
	Replicas int `json:"replicas"`
}

// scaleSwarmService sets a service's replica count, which is how swarm mode
// does both "stop" (zero) and "start" (its usual count).
func (s *Server) scaleSwarmService(w http.ResponseWriter, r *http.Request) error {
	var input scaleSwarmServiceBody
	if !httpapi.DecodeJSON(w, r, &input) {
		return nil
	}
	if input.Replicas < 0 {
		return httpapi.Invalid(w, "Invalid request body",
			[]httpapi.ValidationIssue{{Message: "replicas can't be negative", Path: []string{"replicas"}}})
	}
	err := s.docker.ScaleSwarmService(r.Context(), chi.URLParam(r, "id"), input.Replicas)
	if errors.Is(err, dockerapi.ErrNotFound) {
		return httpapi.Error(w, http.StatusNotFound, "No such swarm service.")
	}
	if err != nil {
		return err
	}
	return httpapi.Answer(w, http.StatusOK, map[string]bool{"ok": true})
}

// restartSwarmService rolls a service's tasks without changing its spec.
func (s *Server) restartSwarmService(w http.ResponseWriter, r *http.Request) error {
	err := s.docker.RestartSwarmService(r.Context(), chi.URLParam(r, "id"))
	if errors.Is(err, dockerapi.ErrNotFound) {
		return httpapi.Error(w, http.StatusNotFound, "No such swarm service.")
	}
	if err != nil {
		return err
	}
	return httpapi.Answer(w, http.StatusOK, map[string]bool{"ok": true})
}

// removeSwarmService removes a swarm service. One already gone isn't an error.
func (s *Server) removeSwarmService(w http.ResponseWriter, r *http.Request) error {
	if err := s.docker.RemoveSwarmService(r.Context(), chi.URLParam(r, "id")); err != nil {
		return err
	}
	return httpapi.Answer(w, http.StatusOK, map[string]bool{"ok": true})
}
